src/groupeddataframe/grouping.jl

```diff
@@ -115,96 +114,204 @@ Base.getindex(gd::GroupedDataFrame, I::AbstractArray{Bool}) =
 
 ```julia
-combine(ga::GroupApplied)
+map(f::Function, gd::GroupedDataFrame)
 ```
 
 ### Arguments
 
-* `ga` : a GroupApplied
+* `gd` : a `GroupedDataFrame` object
 
 ### Returns
 
 * `::DataFrame`
 
+### Details
+
+For each group in `gd`, `f` is passed a `SubDataFrame` view with the corresponding rows.
+`f` can return a single value, a named tuple, a vector, or a data frame.
+This determines the shape of the resulting data frame:
+- A single value gives a data frame with a single column and one row per group.
+- A named tuple gives a data frame with one column for each field in the named tuple
+  and one row per group.
+- A vector gives a data frame with a single column and as many rows
+  for each group as the length of the returned vector for that group.
+- A data frame gives a data frame with the same columns and as many rows
+  for each group as the rows returned for that group.
+
+In all cases, the resulting data frame contains all the grouping columns in addition
+to those listed above. Note that `f` must always return the same type of object for
+all groups, and (if a named tuple or data frame) with the same fields or columns in the
+same order. Returning a single value or a named tuple is significantly faster than
+returning a vector or a data frame.
+
+The resulting data frame will be sorted if `sort=true` was passed to the [`groupby`](@ref)
+call from which `gd` was constructed. Otherwise, ordering of rows is undefined.
+
 ### Examples
 
 ```julia
 df = DataFrame(a = repeat([1, 2, 3, 4], outer=[2]),
                b = repeat([2, 1], outer=[4]),
                c = randn(8))
 gd = groupby(df, :a)
-combine(map(d -> mean(skipmissing(d[:c])), gd))
+map(d -> sum(skipmissing(d[:c])), gd)
 ```
 
+### See also
+
+`by(f, df, cols)` is a shorthand for `map(f, groupby(df, cols))`.
+
 """
-function combine(ga::GroupApplied)
-    gd, vals = ga.gd, ga.vals
-    valscat = _vcat(vals)
-    idx = Vector{Int}(undef, size(valscat, 1))
-    j = 0
-    @inbounds for (start, val) in zip(gd.starts, vals)
-        n = size(val, 1)
-        idx[j .+ (1:n)] .= gd.idx[start]
-        j += n
+function Base.map(f::Function, gd::GroupedDataFrame)
+    if length(gd) > 0
+        idx, valscat = _combine(wrap(f(gd[1])), f, gd)
+        return hcat!(gd.parent[idx, gd.cols], valscat)
+    else
+        return similar(gd.parent, 0)[gd.cols]
     end
-    hcat!(gd.parent[idx, gd.cols], valscat)
 end
 
+function _combine(first::NamedTuple, f::Function, gd::GroupedDataFrame)
+    m = length(first)
+    n = length(gd)
+    idx = Vector{Int}(undef, n)
+    initialcols = ntuple(i -> Vector{typeof(first[i])}(undef, n), m)
+    cols = _combine!(first, initialcols, idx, 1, 1, f, gd, propertynames(first))
+    valscat = DataFrame(collect(cols), collect(propertynames(first)))
+    idx, valscat
+end
+
+function _combine(first::AbstractDataFrame, f::Function, gd::GroupedDataFrame)
+    m = size(first, 2)
+    idx = Vector{Int}()
```

Comment:
`Int[]`?

Reply:
I'd rather keep it that way for consistency with the method above.